Support OPTIONS and HEAD requests for any given route

weierophinney opened the pull request:
This patch accomplishes two goals:

- Updates the `FastRouteRouter` to use the new `RouteResult::fromRoute()` method when creating a successful route result. This ensures that consumers can retrieve the route that matched, along with its path, allowed methods, options, etc.
- Updates the `FastRouteRouter` to always honor `HEAD` and `OPTIONS` requests if the path matches _any_ attached route.

This latter was done to address zendframework/zend-expressive#398 for this routing implementation.

The approach taken updates the `match()` logic such that a routing failure for a `HEAD` or `OPTIONS` request will now have the router loop through all standard HTTP methods in an attempt to dispatch the path; if any match, a successful result is marshaled. This has a potential performance implication if there are many discrete routes with the same path. As such, we will need to recommend that consumers always register a path using `GET`, and, ideally, also register it with `HEAD` and `OPTIONS` (or specify no HTTP methods, which implies `GET`, `HEAD`, and `OPTIONS`). (Since `HEAD` requests will return true if the path matches against `GET`, this approach is fairly performant.)

weierophinney committed:
Use RouteResult::fromRoute() to create result instance

Ensures `Route` instance is composed.

Required updating to zend-expressive-router 1.3.1. One interesting
tidbit: many tests break when HEAD and OPTIONS are implicitly added.
Since FastRoute largely deals with these already, we can remove those
methods from the allowed methods when registering a route with the
underlying `RouteCollector` if they were implicitly added previously.

weierophinney committed:
Support HEAD and OPTIONS requests for any route

This patch solves zendframework/zend-expressive#398 for the FastRoute
implementation, by doing the following:

- Strips `HEAD` and `OPTIONS` from the allowed methods if they were
  added _implicitly_.
- Updates the `match()` logic such that if a `HEAD` or `OPTIONS` request
  fails to match, the router than attempts to match the same path
  against any potential HTTP method. If any succeeds, it returns the
  match result; otherwise, it continues to return the match failure.

In order to enable this, several constants were added to
FastRouteRouter, representing:

- Methods with which to register a route if no methods were provided
  (`GET`, `HEAD`, and `OPTIONS`)
- Methods representing HTTP introspection methods (`GET`, `OPTIONS`)
- The set of all supported HTTP methods by FastRoute

This approach has potential performance overhead, as it will loop
through each of the supported HTTP methods in an attempt to match a HEAD
or OPTIONS request for a given path.

Users should do the following:

- Ensure there is a GET route for a given path; GET implies HEAD support
  in FastRoute, and is the first HTTP method tried after HEAD.
- Alternately, ensure that the first route registerd for a given path
  also registers the HEAD and OPTIONS methods.

michalbundyra commented on src/FastRouteRouter.php:
```diff
@@ -405,11 +440,21 @@ private function injectRoute(Route $route)
+        }
+
+        // Remove HEAD from allowed methods if it was implicitly registered
+        if (in_array(RequestMethod::METHOD_HEAD, $methods, true) && $route->implicitHead()) {
```
I'd change order of these conditions, because `implicitHead()` is simply getter, and `in_array` needs to find element. Isn't it gonna be _more_ efficient?

weierophinney replied:
There's no reason to remove the method if it's not present in the first place. However, it's a moot point, as auto-adding the HEAD and OPTIONS methods when creating the `Route` instance was causing problems with route duplication in the `Application` instance. As such, this check is no longer needed; we'll check for `implicitHead()` and/or `implicitOptions()` in the dispatcher middleware instead. I've now removed these conditionals.

weierophinney committed:
Remove checks for implicit*() when injecting routes

No longer necessary, as zend-expressive-router 1.3.2 no longer
auto-injects the HEAD and OPTIONS methods.

weierophinney committed:
Renamed INTROSPECT to IMPLICIT

- Better indicates purpose
- Matches patches against other router implementations
